crates/adapters/src/format/json/input.rs

```diff
@@ -291,152 +327,280 @@ impl JsonParser {
-                    {
-                        let updates = updates.map_err(|e| {
-                            let json_str = String::from_utf8_lossy(bytes);
-                            anyhow!("error deserializing string as a JSON array of updates: {e}\nInvalid JSON: {json_str}\nExample valid JSON: '[{{\"insert\": {{...}} }}, {{\"delete\": {{...}} }}]'")
-                        })?;
-                        for update in updates.into_iter() {
-                            num_updates += self.apply_insdel_update(update)?;
-                        }
-                    }
-                } else {
-                    for update in
-                        serde_json::Deserializer::from_slice(bytes).into_iter::<InsDelUpdate>()
-                    {
-                        // println!("update: {update:?}");
-                        let update = update.map_err(|e| {
-                            let json_str = String::from_utf8_lossy(bytes);
-                            anyhow!("error deserializing string as a single-row update: {e}\nInvalid JSON: {json_str}\nExample valid JSON: '{{\"insert\": {{...}} }}'")
-                        })?;
-                        num_updates += self.apply_insdel_update(update)?;
+        let mut stream = serde_json::Deserializer::from_slice(bytes).into_iter::<&RawValue>();
+
+        while let Some(update) = stream.next() {
+            let update = match update {
+                Err(e) => {
+                    let json_str = String::from_utf8_lossy(&bytes[stream.byte_offset()..]);
+                    errors.push(ParseError::text_envelope_error(
+                        format!("failed to parse string as a JSON document: {e}"),
+                        &json_str,
+                        None,
+                    ));
+                    if !self.config.array {
+                        self.flush();
                     }
+                    return (num_updates, errors);
                 }
-            }
-            JsonUpdateFormat::Debezium => {
-                if self.config.array {
-                    for updates in serde_json::Deserializer::from_slice(bytes)
-                        .into_iter::<Vec<DebeziumUpdate>>()
-                    {
-                        let updates = updates.map_err(|e| {
-                            let json_str = String::from_utf8_lossy(bytes);
-                            anyhow!("error deserializing string as a JSON array of Debezium CDC records: {e}\nInvalid JSON: {json_str}\nExample valid JSON: '[{{\"payload\": {{\"op\": \"u\", \"before\": {{...}}, \"after\": {{...}} }} }}]'")
-                        })?;
-                        for update in updates.into_iter() {
-                            num_updates += self.apply_debezium_update(update)?;
+                Ok(update) => update,
+            };
+
+            match self.config.update_format {
+                JsonUpdateFormat::InsertDelete => {
```

Comment:
should each of these cases be a separate function?

Reply:
I will refactor this using traits in the next PR.